### PR TITLE
Cxx11

### PR DESCRIPTION
**Pull request by lhmouse**
This commit adds reference implementations for C++11 threading facilities:

* `once_flag`
* `mutex` and `timed_mutex`
* `recursive_mutex` and `recursive_timed_mutex`
* `condition_variable`
* `thread`
* `thread_specific_ptr` (_non-standard, inspired by boost_)

This closes #77.

### CODE REVIEW

File: src/cxx11.hpp [hunk 1]
```diff
@@ -0,0 +1,854 @@
...
+// At the moment, all results are discarded, so the expression always has a
+// type of `void` for simplicity.
+#if defined(__cpp_lib_invoke)
+
+#  define __MCF_VOID_INVOKE(_F, ...)   (::std::invoke(_F, __VA_ARGS__), void())
+
+#else  // std::invoke
+
+template<class _Member, class _Class, class... _Args>
+__MCF_CXX14(constexpr)
```
[the remark this answers is not in the document]
Comment (lhmouse) [reply src/cxx11.hpp]: Really. I removed that.

File: src/cxx11.hpp [hunk 2]
```diff
@@ -0,0 +1,865 @@
...
+
+    int __err = ::__MCF_cxa_thread_atexit((__MCF_cxa_dtor_cdecl*)(intptr_t) __target,
+                                          __ptr, &__dso_handle);
+    if(__err != 0)
+      __MCF_THROW_SYSTEM_ERROR(not_enough_memory, "__MCF_cxa_thread_atexit");
+  }
+
+// Create a thread. The C++ standard requires that arguments passed to the
+// constructor of `std::thread` be passed to the thread function object as
+// decay-copied rvalues. This is the object that stores copies of them and
```
Comment (FrankHB) [inline src/cxx11.hpp:233]: Probably better "stores" -> "saves"/"holds".

File: src/cxx11.hpp [hunk 3]
```diff
@@ -0,0 +1,865 @@
...
+      }
+
+    void
+    unlock() noexcept  // strengthened
+      {
+        ::_MCF_mutex_unlock(this->_M_mtx);
+      }
+  };
+
+// Reference implementation for [thread.timedmutex.recursive]
```
Comment (FrankHB) [inline src/cxx11.hpp:374]: We need no emphasize on "reference implementation" more than once, probably.

File: src/cxx11.hpp [hunk 4]
```diff
@@ -0,0 +1,865 @@
...
+            ::_MCF_event _M_event[1];
+            _My_invoker _M_invoker[1];
+
+            static
+            void
+            __thunk(::_MCF_thread* __thr)
+              {
+                _My_data* __my = (_My_data*) ::_MCF_thread_get_data(__thr);
+
+                // Check whether `_M_invoker` has been initialized. If its
```
Comment (FrankHB) [inline src/cxx11.hpp:601]: Whose constructor? `_M_invoker` is of an array type. I can hardly get the point.

File: src/cxx11.hpp [hunk 5]
```diff
@@ -0,0 +1,865 @@
...
+        return *this;
+      }
+
+    ~thread() noexcept
+      {
+        if(this->_M_thr)
+          ::std::terminate();
+      }
+
+    using id = uint32_t;
```
Comment (FrankHB) [inline src/cxx11.hpp:662]: This is not conforming, because the standard requires it to be a class type.

File: src/cxx11.hpp (hunk 4 above)
Comment (lhmouse) [reply src/cxx11.hpp:605]: This had better be `*_M_invoker`.

File: src/cxx11.hpp (hunk 5 above)
Comment (lhmouse) [reply src/cxx11.hpp]: well, what about `std::hash`?

**Comment by lhmouse** [comment]
@frankhb We need some testcases at the moment.

**Comment by lhmouse** [comment]
Ready to merge?

**Comment by FrankHB** [comment]
LGTM.

**Pull request merged**